rpc/client/eventstream/eventstream_test.go

```diff
@@ -0,0 +1,226 @@
+package eventstream_test
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/fortytw2/leaktest"
+	"github.com/google/go-cmp/cmp"
+
+	"github.com/tendermint/tendermint/internal/eventlog"
+	"github.com/tendermint/tendermint/internal/eventlog/cursor"
+	"github.com/tendermint/tendermint/rpc/client/eventstream"
+	rpccore "github.com/tendermint/tendermint/rpc/core"
+	coretypes "github.com/tendermint/tendermint/rpc/core/types"
+	"github.com/tendermint/tendermint/types"
+)
+
+func TestStream_filterOrder(t *testing.T) {
+	defer leaktest.Check(t)
+
+	s := newStreamTester(t, `tm.event = 'good'`, eventlog.LogSettings{
+		WindowSize: 30 * time.Second,
+	}, nil)
+
+	// Verify that events are delivered in forward time order (i.e., that the
+	// stream unpacks the pages correctly) and that events not matching the
+	// query (here, type="bad") are skipped.
+	//
+	// The minimum batch size is 16 and half the events we publish match, so we
+	// publish > 32 items (> 16 good) to ensure we exercise paging.
+	etype := [2]string{"good", "bad"}
+	var items []testItem
+	for i := 0; i < 40; i++ {
+		s.advance(100 * time.Millisecond)
+		text := fmt.Sprintf("item%d", i)
+		cur := s.publish(etype[i%2], text)
+
+		// Even-numbered items match the target type.
+		if i%2 == 0 {
+			items = append(items, makeTestItem(cur, text))
+		}
+	}
+
+	s.start()
+	for _, itm := range items {
+		s.mustItem(t, itm)
+	}
+	s.stopWait()
+}
+
+func TestStream_lostItem(t *testing.T) {
+	defer leaktest.Check(t)
+
+	s := newStreamTester(t, ``, eventlog.LogSettings{
+		WindowSize: 30 * time.Second,
+	}, nil)
+
+	// Publish an item and let the client observe it.
+	cur := s.publish("ok", "whatever")
+	s.start()
+	s.mustItem(t, makeTestItem(cur, "whatever"))
+	s.stopWait()
+
+	// Time passes, and cur expires out of the window.
+	s.advance(50 * time.Second)
+	next1 := s.publish("ok", "more stuff")
+	s.advance(15 * time.Second)
+	next2 := s.publish("ok", "still more stuff")
+
+	// At this point, the oldest item in the log is newer than the point at
+	// which we continued, we should get an error.
+	s.start()
+	var missed *eventstream.MissedItemsError
+	if err := s.mustError(t); !errors.As(err, &missed) {
+		t.Errorf("Wrong error: got %v, want %T", err, missed)
+	} else {
+		t.Logf("Correctly reported missed item: %v", missed)
+	}
+
+	// If we reset the stream and continue from head, we should catch up.
+	s.stopWait()
+	s.stream.Reset()
+	s.start()
+
+	s.mustItem(t, makeTestItem(next1, "more stuff"))
+	s.mustItem(t, makeTestItem(next2, "still more stuff"))
+	s.stopWait()
+}
+
+// testItem is a wrapper for comparing item results in a friendly output format
+// for the cmp package.
+type testItem struct {
+	Cursor string
+	Data   string
+
+	// N.B. Fields exported to simplify use in cmp.
+}
+
+func makeTestItem(cur, data string) testItem {
+	return testItem{
+		Cursor: cur,
+		Data:   fmt.Sprintf(`%q`, data),
```

Comment:
The original _Sprintf_ formatting is:
```
`{"type":%q,"value":%q}`, types.EventDataString("").TypeTag(), data
```
Any reason why you simplified it?

Reply:
To adjust [json.Marshal](https://github.com/tendermint/tendermint/blob/cca48d01ae76772b70e5c733c3696dce544bedfd/rpc/core/events.go#L276) while origin could make use of [jsontypes.Marshal](https://github.com/tendermint/tendermint/blob/ab32f5a9b6b3f3ded49a86989516898b61065a52/internal/rpc/core/events.go#L253)

Reply:
I see, since #7600 hasn't been backported, I believe.
If that is the case, I'll approve.
In any case, do you think if would be beneficial to backport #7600 in a separate PR?

Reply:
I think the [tradeoff](https://github.com/tendermint/tendermint/pull/9470#discussion_r977102153) is mentioned before, and only 3 small PRs left for this [issue](https://github.com/tendermint/tendermint/issues/9305).